src/software/new_geom/rectangle.cpp

```diff
@@ -0,0 +1,97 @@
+#include "software/new_geom/rectangle.h"
+
+Rectangle::Rectangle(const Point &point1, const Point &point2)
+    : Polygon({Point(point1.x() < point2.x() ? point1.x() : point2.x(),
+                     point1.y() < point2.y() ? point1.y() : point2.y()),
+               Point(point1.x() < point2.x() ? point1.x() : point2.x(),
+                     point1.y() > point2.y() ? point1.y() : point2.y()),
+               Point(point1.x() > point2.x() ? point1.x() : point2.x(),
+                     point1.y() > point2.y() ? point1.y() : point2.y()),
+               Point(point1.x() > point2.x() ? point1.x() : point2.x(),
+                     point1.y() < point2.y() ? point1.y() : point2.y())})
+{
+}
+
+Vector Rectangle::diagonal() const
+{
+    return points_[2] - points_[0];
+}
+
+double Rectangle::xLength() const
+{
+    return diagonal().x();
+}
+
+double Rectangle::yLength() const
+{
+    return diagonal().y();
+}
+
+double Rectangle::area() const
+{
+    return diagonal().x() * diagonal().y();
+}
+
+Point Rectangle::centre() const
+{
+    return Point(negXNegYCorner() + (diagonal() / 2));
+}
+
+Point Rectangle::posXPosYCorner() const
+{
+    return negXNegYCorner() + diagonal();
+}
+
+Point Rectangle::negXPosYCorner() const
+{
+    return negXNegYCorner() + Vector(0, diagonal().y());
+}
```

Comment:
For each of the "corner" functions here, could we just return a specific point from `points_` since we enforce the order in the constructor?

Reply:
Fixed in latest update